internal/dataplane/sendconfig/backoff_strategy.go

```diff
@@ -0,0 +1,80 @@
+package sendconfig
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/adminapi"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/metrics"
+)
+
+type ErrUpdateSkippedDueToBackoffStrategy struct {
+	explanation string
+}
+
+func NewErrUpdateSkippedDueToBackoffStrategy(explanation string) ErrUpdateSkippedDueToBackoffStrategy {
+	return ErrUpdateSkippedDueToBackoffStrategy{explanation: explanation}
+}
+
+func (e ErrUpdateSkippedDueToBackoffStrategy) Error() string {
+	return fmt.Sprintf("update skipped due to a backoff strategy not being satisfied: %s", e.explanation)
+}
+
+func (e ErrUpdateSkippedDueToBackoffStrategy) Is(err error) bool {
+	return errors.Is(err, ErrUpdateSkippedDueToBackoffStrategy{})
+}
+
+// UpdateStrategyWithBackoff decorates any UpdateStrategy to respect a passed adminapi.UpdateBackoffStrategy.
+type UpdateStrategyWithBackoff struct {
+	decorated       UpdateStrategy
+	backoffStrategy adminapi.UpdateBackoffStrategy
+	log             logrus.FieldLogger
+}
+
+func NewUpdateStrategyWithBackoff(
+	decorated UpdateStrategy,
+	backoffStrategy adminapi.UpdateBackoffStrategy,
+	log logrus.FieldLogger,
+) UpdateStrategyWithBackoff {
+	return UpdateStrategyWithBackoff{
+		decorated:       decorated,
+		backoffStrategy: backoffStrategy,
+		log:             log,
+	}
+}
+
+// Update will ensure that the decorated UpdateStrategy.Update is called only when an underlying
+// UpdateBackoffStrategy.CanUpdate is satisfied.
+// In case it's not, it will return a predefined ErrUpdateSkippedDueToBackoffStrategy.
+// In case it is, apart from calling UpdateStrategy.Update, it will also register a success or a failure of an update
+// attempt so that the UpdateBackoffStrategy can keep track of it.
+func (s UpdateStrategyWithBackoff) Update(ctx context.Context, targetContent ContentWithHash) (
+	err error,
+	resourceErrors []ResourceError,
+	resourceErrorsParseErr error,
+) {
+	if canUpdate, whyNot := s.backoffStrategy.CanUpdate(targetContent.Hash); !canUpdate {
+		return NewErrUpdateSkippedDueToBackoffStrategy(whyNot), nil, nil
+	}
+
+	err, resourceErrors, resourceErrorsParseErr = s.decorated.Update(ctx, targetContent)
+	if err != nil {
+		s.log.WithError(err).Debug("Update failed, registering it for backoff strategy")
+		s.backoffStrategy.RegisterUpdateFailure(err, targetContent.Hash)
+	} else {
+		s.backoffStrategy.RegisterUpdateSuccess()
+	}
+
+	return err, resourceErrors, resourceErrorsParseErr
+}
+
+func (s UpdateStrategyWithBackoff) MetricsProtocol() metrics.Protocol {
+	return s.decorated.MetricsProtocol()
+}
+
+func (s UpdateStrategyWithBackoff) Type() string {
+	return fmt.Sprintf("WithBackoff(%s)", s.decorated.Type())
+}
```

Comment:
Take it or leave it comment:

I'm not particularly fond of using "type as string" if it's not strictly being used for logging/printing purposes.

The only purpose in this PR that I see for this is to distinguish the types of update strategy in tests which I believe might be a little bit of a stretch. While I'm not sure there's a better solution out there, 1 possible approach I can think of is to make `UpdateStrategyWithBackoff` to have a type parameter and then use that within the assertions in tests instead of using string comparison.

WDYT?

Reply:
I get your concern 👍 I liked your proposal of using a type parameter, and I tried going with it but unfortunately this doesn't help as `NewUpdateStrategyWithBackoff` is called in a context in which we don't have a concrete type but just an `UpdateStrategy` interface. So instead of `UpdateStrategyWithBackoff[UpdateStrategyDBMode]` we end up with `UpdateStrategyWIthBackoff[UpdateStrategy]` being produced which is not useful in test assertions. 

How about a compromise of using the output of `Type()` as a logger field? :) 410b6fd